File: frontend/lib/shared/models/reservation.dart
/// Modèle de réservation unifié
class Reservation {
  final String id;
  final DateTime date;
  final String time;
  final int duration; // en minutes
  final int partySize;
  final String status;
  final String? notes;
  final String? specialRequests;
  final String? clientName;
  final String? clientEmail;
  final String? clientPhone;
  final String? managementToken;
  final bool requiresPayment;
  final double? depositAmount;
  final String paymentStatus;
  final String restaurantId;
  final String? tableId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Reservation({
    required this.id,
    required this.date,
    required this.time,
    required this.duration,
    required this.partySize,
    required this.status,
    this.notes,
    this.specialRequests,
    this.clientName,
    this.clientEmail,
    this.clientPhone,
    this.managementToken,
    this.requiresPayment = false,
    this.depositAmount,
    this.paymentStatus = 'NONE',
    required this.restaurantId,
    this.tableId,
    this.createdAt,
    this.updatedAt,
  });

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      id: json['id'] as String,
      date: DateTime.parse(json['date'] as String),
      time: json['time'] as String,
      duration: json['duration'] as int,
      partySize: json['partySize'] as int,
      status: json['status'] as String,
      notes: json['notes'] as String?,
      specialRequests: json['specialRequests'] as String?,
      clientName: json['clientName'] as String?,
      clientEmail: json['clientEmail'] as String?,
      clientPhone: json['clientPhone'] as String?,
      managementToken: json['managementToken'] as String?,
      requiresPayment: json['requiresPayment'] as bool? ?? false,
      depositAmount: (json['depositAmount'] as num?)?.toDouble(),
      paymentStatus: json['paymentStatus'] as String? ?? 'NONE',
      restaurantId: json['restaurantId'] as String,
      tableId: json['tableId'] as String?,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'date': date.toIso8601String(),
      'time': time,
      'duration': duration,
      'partySize': partySize,
      'status': status,
      'notes': notes,
      'specialRequests': specialRequests,
      'clientName': clientName,
      'clientEmail': clientEmail,
      'clientPhone': clientPhone,
      'managementToken': managementToken,
      'requiresPayment': requiresPayment,
      'depositAmount': depositAmount,
      'paymentStatus': paymentStatus,
      'restaurantId': restaurantId,
      'tableId': tableId,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  Reservation copyWith({
    String? id,
    DateTime? date,
    String? time,
    int? duration,
    int? partySize,
    String? status,
    String? notes,
    String? specialRequests,
    String? clientName,
    String? clientEmail,
    String? clientPhone,
    String? managementToken,
    bool? requiresPayment,
    double? depositAmount,
    String? paymentStatus,
    String? restaurantId,
    String? tableId,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Reservation(
      id: id ?? this.id,
      date: date ?? this.date,
      time: time ?? this.time,
      duration: duration ?? this.duration,
      partySize: partySize ?? this.partySize,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      specialRequests: specialRequests ?? this.specialRequests,
      clientName: clientName ?? this.clientName,
      clientEmail: clientEmail ?? this.clientEmail,
      clientPhone: clientPhone ?? this.clientPhone,
      managementToken: managementToken ?? this.managementToken,
      requiresPayment: requiresPayment ?? this.requiresPayment,
      depositAmount: depositAmount ?? this.depositAmount,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      restaurantId: restaurantId ?? this.restaurantId,
      tableId: tableId ?? this.tableId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// Statuts de réservation
enum ReservationStatus {
  PENDING,
  CONFIRMED,
  CANCELLED,
  COMPLETED,
  NO_SHOW,
}

/// Statuts de paiement
enum PaymentStatus {
  NONE,
  PENDING,
  COMPLETED,
  FAILED,
  REFUNDED,
  PARTIALLY_REFUNDED,
}

File: backend/src/domain/entities/Reservation.ts
import { ReservationStatus, PaymentStatus } from '@prisma/client';

interface ReservationProps {
  id: string;
  date: Date;
  time: string;
  duration: number;
  partySize: number;
  status: ReservationStatus;
  requiresPayment: boolean;
  depositAmount: number;
  paymentStatus: PaymentStatus;
  notes?: string;
  specialRequests?: string;
  clientName?: string;
  clientEmail?: string;
  clientPhone?: string;
  managementToken?: string;
  tokenExpiresAt?: Date;
  cancellationReason?: string;
  userId?: string;
  restaurantId: string;
  tableId?: string;
  createdAt: Date;
  updatedAt: Date;
}

export class Reservation {
  private props: ReservationProps;

  constructor(props: ReservationProps) {
    this.props = props;
  }

  // Getters
  get id(): string { return this.props.id; }
  get date(): Date { return this.props.date; }
  get time(): string { return this.props.time; }
  get duration(): number { return this.props.duration; }
  get partySize(): number { return this.props.partySize; }
  get status(): ReservationStatus { return this.props.status; }
  get requiresPayment(): boolean { return this.props.requiresPayment; }
  get depositAmount(): number { return this.props.depositAmount; }
  get paymentStatus(): PaymentStatus { return this.props.paymentStatus; }
  get notes(): string | undefined { return this.props.notes; }
  get specialRequests(): string | undefined { return this.props.specialRequests; }
  get clientName(): string | undefined { return this.props.clientName; }
  get clientEmail(): string | undefined { return this.props.clientEmail; }
  get clientPhone(): string | undefined { return this.props.clientPhone; }
  get managementToken(): string | undefined { return this.props.managementToken; }
  get tokenExpiresAt(): Date | undefined { return this.props.tokenExpiresAt; }
  get cancellationReason(): string | undefined { return this.props.cancellationReason; }
  get userId(): string | undefined { return this.props.userId; }
  get restaurantId(): string { return this.props.restaurantId; }
  get tableId(): string | undefined { return this.props.tableId; }
  get createdAt(): Date { return this.props.createdAt; }
  get updatedAt(): Date { return this.props.updatedAt; }

  // Business logic methods
  isGuestReservation(): boolean {
    return !this.props.userId && !!this.props.managementToken;
  }

  isTokenExpired(): boolean {
    if (!this.props.tokenExpiresAt) return false;
    return this.props.tokenExpiresAt < new Date();
  }

  canBeCancelled(): boolean {
    return this.props.status === ReservationStatus.PENDING || 
           this.props.status === ReservationStatus.CONFIRMED;
  }

  canBeModified(): boolean {
    return this.props.status === ReservationStatus.PENDING || 
           this.props.status === ReservationStatus.CONFIRMED;
  }

  isActive(): boolean {
    return this.props.status !== ReservationStatus.CANCELLED;
  }

  // Factory method to create Reservation from Prisma model
  static fromPrisma(prismaReservation: any): Reservation {
    return new Reservation({
      id: prismaReservation.id,
      date: prismaReservation.date,
      time: prismaReservation.time,
      duration: prismaReservation.duration,
      partySize: prismaReservation.partySize,
      status: prismaReservation.status,
      requiresPayment: prismaReservation.requiresPayment,
      depositAmount: prismaReservation.depositAmount,
      paymentStatus: prismaReservation.paymentStatus,
      notes: prismaReservation.notes,
      specialRequests: prismaReservation.specialRequests,
      clientName: prismaReservation.clientName,
      clientEmail: prismaReservation.clientEmail,
      clientPhone: prismaReservation.clientPhone,
      managementToken: prismaReservation.managementToken,
      tokenExpiresAt: prismaReservation.tokenExpiresAt,
      cancellationReason: prismaReservation.cancellationReason,
      userId: prismaReservation.userId,
      restaurantId: prismaReservation.restaurantId,
      tableId: prismaReservation.tableId,
      createdAt: prismaReservation.createdAt,
      updatedAt: prismaReservation.updatedAt,
    });
  }

  toJSON() {
    return {
      id: this.id,
      date: this.date,
      time: this.time,
      duration: this.duration,
      partySize: this.partySize,
      status: this.status,
      requiresPayment: this.requiresPayment,
      depositAmount: this.depositAmount,
      paymentStatus: this.paymentStatus,
      notes: this.notes,
      specialRequests: this.specialRequests,
      clientName: this.clientName,
      clientEmail: this.clientEmail,
      clientPhone: this.clientPhone,
      managementToken: this.managementToken,
      tokenExpiresAt: this.tokenExpiresAt,
      cancellationReason: this.cancellationReason,
      userId: this.userId,
      restaurantId: this.restaurantId,
      tableId: this.tableId,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }
}


